File: house-robber.js
// https://leetcode.com/problems/house-robber/
// https://leetcode.com/submissions/detail/294162482/

/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    let ifRobbedPrevious = 0; 	// max monney can get if rob current house
	let ifDidntRobPrevious = 0; // max money can get if not rob current house
	    
	    // We go through all the values, we maintain two counts, 1) if we rob this cell, 2) if we didn't rob this cell
	    for(let i=0; i < nums.length; i++) 
	    {
	    	// If we rob current cell, previous cell shouldn't be robbed. So, add the current value to previous one.
	        let currRobbed = ifDidntRobPrevious + nums[i];
	        
	        // If we don't rob current cell, then the count should be max of the previous cell robbed and not robbed
	        let currNotRobbed = Math.max(ifDidntRobPrevious, ifRobbedPrevious); 
	        
	        // Update values for the next round
	        ifDidntRobPrevious  = currNotRobbed;
	        ifRobbedPrevious = currRobbed;
	    }
	    
	    return Math.max(ifRobbedPrevious, ifDidntRobPrevious);
};

File: python/Dynamic Programming/house-robber.py
# https://leetcode.com/problems/house-robber/submissions/

class Solution:
    def rob(self, nums: List[int]) -> int:
        
        dp = [0] + [num for num in nums]
        # print(dp)       
        
        for i in range(2, len(nums)+1):
            # dp[i-1] => Last  house was robbed
            # dp[i-2]+nums[i-1] => Last house wasn't robbed
            dp[i] = max(dp[i-1], dp[i-2]+nums[i-1])
            # print("i = ", i, " dp = ", dp)
        
        return dp[-1]